compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -434,20 +435,50 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, '_, 'infcx, 'tcx> {
                 let parent = self.infcx.tcx.parent_hir_node(expr.hir_id);
                 let (def_id, args, offset) = if let hir::Node::Expr(parent_expr) = parent
                     && let hir::ExprKind::MethodCall(_, _, args, _) = parent_expr.kind
-                    && let Some(def_id) = typeck.type_dependent_def_id(parent_expr.hir_id)
                 {
-                    (def_id.as_local(), args, 1)
+                    (typeck.type_dependent_def_id(parent_expr.hir_id), args, 1)
                 } else if let hir::Node::Expr(parent_expr) = parent
                     && let hir::ExprKind::Call(call, args) = parent_expr.kind
                     && let ty::FnDef(def_id, _) = typeck.node_type(call.hir_id).kind()
                 {
-                    (def_id.as_local(), args, 0)
+                    (Some(*def_id), args, 0)
                 } else {
                     (None, &[][..], 0)
                 };
+
+                // If the moved value is a mut reference, it is used in a
+                // generic function and it's type is a generic param, it can be
+                // reborrowed to avoid moving.
+                // for example:
+                // struct Y(u32);
+                // x's type is '& mut Y' and it is used in `fn generic<T>(x: T) {}`.
+                let is_sugg_reborrow = || {
+                    if let Some(def_id) = def_id
+                        && self.infcx.tcx.def_kind(def_id).is_fn_like()
+                        && let Some(pos) = args.iter().position(|arg| arg.hir_id == expr.hir_id)
+                        && let ty::Param(_) =
+                            self.infcx.tcx.fn_sig(def_id).skip_binder().skip_binder().inputs()
+                                [pos + offset]
+                                .kind()
+                    {
+                        let place = &self.move_data.move_paths[mpi].place;
+                        let ty = place.ty(self.body, self.infcx.tcx).ty;
+                        if let ty::Ref(_, _, hir::Mutability::Mut) = ty.kind() {
+                            return true;
+                        }
+                    }
+                    false
+                };
+                *has_suggest_reborrow = is_sugg_reborrow();
```

Comment:
why use a closure here instead of inlining the code below into `return true`?

Reply:
> why use a closure here instead of inlining the code below into `return true`?

Thank you. I just think that because it's in the function `suggest_ref_or_clone function`, this judgment code checks a special situation and returns early, so maybe this check code could be organized into a closure. Should I abandon the closure here?

Reply:
yes, abandon the closure

Reply:
> yes, abandon the closure

Done. Thank you.